test/Chainweb/Test/CutDB.hs

```diff
@@ -196,7 +200,34 @@ extendAwait cdb pact i p = race gen (awaitCut cdb p) >>= \case
     Left _ -> return Nothing
     Right c -> return (Just c)
   where
-    gen = void $! S.effects $ extendTestCutDb cdb pact i
+    gen = void
+        $ S.foldM_ checkCut (return 0) return
+        $ S.map (view (_1 . cutHeight))
+        $ extendTestCutDb cdb pact i
+
+    checkCut prev cur = do
+        unless (prev < cur) $ throwM $ InternalInvariantViolation $ unexpectedMsg
+            "New cut is not larger than the previous one. This is bug in Chainweb.Test.CutDB"
+            (Expected prev)
+            (Actual cur)
+        return cur
+
+-- | Extend cut db via 'extendAwait', retrying n-many times in the case where
+-- no cut fulfils the current condition.
+--
+extendAwaitRetryN
+    :: PayloadCas cas
+    => CutDb cas
+    -> WebPactExecutionService
+    -> Natural
+    -> Natural
+    -> (Cut -> Bool)
+    -> IO Cut
+extendAwaitRetryN _ _ _ 0 _ = throwM $
+    userError "Cut extension retry limit exceeded. This is a bug in Chainweb.Test.CutDb"
+extendAwaitRetryN cdb pact i n p = extendAwait cdb pact i p >>= \case
+    Nothing -> extendAwaitRetryN cdb pact 1 (pred n) p
```

Comment:
@larskuhtz in the case where we don't find any cuts that satisfies our predicate, this function just bumps the extension by 1 block until either we run out of retries (highly unlikely), or a block is found (very likely)

Reply:
Wouldn't this have the exact same effect as calling `extendAwait cab pact (i * n) p`?

Reply:
It's actually `i + n` for some `n` (note the 1 in `extendAwaitRetryN cdb pact 1 (pred n) p`!) . You said we'd want the minimum amount of blocks mined, so it makes sense extend the cut db by 1 block at a time. It seems (and I printed out the logs to double check), that we are only ever behind by 1 or 2 blocks at a time when we receive `Nothing`.

Reply:
That would still be `n (n - 1) / 2`, thus quadratic, wouldn't it?

Reply:
Assuming no bugs, `extendAwait cdb pact n` mines up to `n` blocks, each on a randomly picked chain, until the predicate is satisfied. It seems to me that `extendAwaitRetryN cdb pact 1 (pred n)` should have the same effect as `extendAwait cdb pact (n + 1)`. If both have different semantics, it would mean that `extendAwait` is buggy.